lib/Mojo/mysql.pm

```diff
@@ -27,62 +27,48 @@ sub db {
   delete @$self{qw(pid queue)} unless ($self->{pid} //= $$) eq $$;
 
   my ($dbh, $handle) = @{$self->_dequeue};
-  return Mojo::mysql::Database->new(dbh => $dbh, handle => $handle, mysql => $self);
+  my $db = ($self->options->{use_dbi} // 1) ?
+    Mojo::mysql::DBI::Database->new(dbh => $dbh, handle => $handle, mysql => $self) :
+    Mojo::mysql::Native::Database->new(connection => $dbh, mysql => $self);
+
+  if (!$dbh) {
+    $db->connect(map { $self->$_ } qw(url username password options));
+    $self->emit(connection => $db);
+  }
+  return $db;
 }
 
 sub from_string {
   my ($self, $str) = @_;
 
-  # Protocol
-  return $self unless $str;
-  my $url = Mojo::URL->new($str);
-  croak qq{Invalid MySQL connection string "$str"} unless $url->protocol eq 'mysql';
-
-  # Database
-  my $dsn = 'dbi:mysql:dbname=' . $url->path->parts->[0];
-
-  # Host and port
-  if (my $host = $url->host) { $dsn .= ";host=$host" }
-  if (my $port = $url->port) { $dsn .= ";port=$port" }
+  my $parts = parse_url($str);
+  croak qq{Invalid MySQL connection string "$str"} unless defined $parts;
 
-  # Username and password
-  if (($url->userinfo // '') =~ /^([^:]+)(?::([^:]+))?$/) {
-    $self->username($1);
-    $self->password($2) if defined $2;
-  }
+  # Only for Compatibility
+  $self->username($parts->{username}) if exists $parts->{username};
+  $self->password($parts->{password}) if exists $parts->{password};
+  $self->dsn($parts->{dsn});
+  @{$self->options}{keys %{$parts->{options}}} = values %{$parts->{options}};
 
-  # Options
-  my $hash = $url->query->to_hash;
-  @{$self->options}{keys %$hash} = values %$hash;
+  load_class(
+    ($self->options->{use_dbi} // 1) ? 'Mojo::mysql::DBI::Database' : 'Mojo::mysql::Native::Database');
 
-  return $self->dsn($dsn);
+  return $self->url($str);
 }
 
 sub new { @_ > 1 ? shift->SUPER::new->from_string(@_) : shift->SUPER::new }
 
 sub _dequeue {
   my $self = shift;
-  my $dbh;
 
   while (my $c = shift @{$self->{queue} || []}) { return $c if $c->[0]->ping }
-  $dbh = DBI->connect(map { $self->$_ } qw(dsn username password options));
-
-  # <mst> batman's probably going to have more "fun" than you have ...
-  # especially once he discovers that DBD::mysql randomly reconnects under
-  # you, silently, but only if certain env vars are set
-  # hint: force-set mysql_auto_reconnect or whatever it's called to 0
-  $dbh->{mysql_auto_reconnect} = 0;
-  # Maintain Commits with Mojo::mysql::Transaction
-  $dbh->{AutoCommit} = 1;
-
-  $self->emit(connection => $dbh);
-  [$dbh];
+  return [undef, undef];
```

Comment:
**_enqueue** returns undef if there is no available handle to reuse.
**connect** called in **db** method if **_enqueue** returns undef.